File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_18.py
# E
# DE
# CDE
# BCDE
# ABCDE

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
start = 'E'
n = 5
for i in range(1, n + 1):
    line = ""
    for j in range(1, i + 1):
        line += chr(ord(start) - i + j)
    print(line)

#----------------------------------
# BETTER APPROACH (String multiplication)
#----------------------------------
start = 'E'
n = 5
for i in range(1, n + 1):
    row = "".join(chr(ord(start) - i + j) for j in range(1, i + 1))
    print(row)

#----------------------------------
# BEST APPROACH (List comprehension)
#----------------------------------
start = 'E'
n = 5
print("\n".join("".join(chr(ord(start) - i + j) for j in range(1, i + 1)) for i in range(1, n + 1)))

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_18.cpp
// E
// DE
// CDE
// BCDE
// ABCDE

//----------------------------------
// BRUTE FORCE APPROACH (Nested loops)
//----------------------------------
#include <iostream>
using namespace std;

int main()
{
    char start = 'E';
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << char(start - i + j);
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String constructor)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    char start = 'E';
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        string row;
        for (int j = 1; j <= i; j++)
        {
            row += char(start - i + j);
        }
        cout << row << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Single pre-allocated string)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    char start = 'E';
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        string row(i, ' ');
        for (int j = 1; j <= i; j++)
        {
            row[j - 1] = char(start - i + j);
        }
        cout << row << endl;
    }
    return 0;
}